File: pages/refactorings/extract-method/03-expression-eval/Before.scala
// Evaluate an arithmetic expression; division by zero has no value.
object Before:
  enum Expr:
    case Lit(n: Int)
    case Add(l: Expr, r: Expr)
    case Mul(l: Expr, r: Expr)
    case Div(l: Expr, r: Expr)
  import Expr.*

  def eval(e: Expr): Option[Int] = e match
    case Lit(n)    => Some(n)
    case Add(l, r) => for a <- eval(l); b <- eval(r) yield a + b
    case Mul(l, r) => for a <- eval(l); b <- eval(r) yield a * b
    case Div(l, r) => for a <- eval(l); b <- eval(r); if b != 0 yield a / b

File: pages/refactorings/extract-method/03-expression-eval/Before.hs
-- Evaluate an arithmetic expression; division by zero has no value.
module Before where

data Expr = Lit Int | Add Expr Expr | Mul Expr Expr | Div Expr Expr
  deriving Show

eval :: Expr -> Maybe Int
eval (Lit n)   = Just n
eval (Add l r) = do
  a <- eval l
  b <- eval r
  Just (a + b)
eval (Mul l r) = do
  a <- eval l
  b <- eval r
  Just (a * b)
eval (Div l r) = do
  a <- eval l
  b <- eval r
  if b == 0 then Nothing else Just (a `div` b)
